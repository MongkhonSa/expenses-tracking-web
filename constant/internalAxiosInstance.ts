import axios from "axios";
import { resHandler, internalErrorHandler } from "./handler";

const internalAxiosInstance = axios.create({
  baseURL: "http://localhost:3000/api",
});

internalAxiosInstance.interceptors.response.use(
  resHandler,
  internalErrorHandler
);
export default internalAxiosInstance;
