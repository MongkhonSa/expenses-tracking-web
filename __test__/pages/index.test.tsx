import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import internalAxiosInstance from "../../constant/internalAxiosInstance";
import HomePage, { Home } from "../../pages/index";
import "jest-styled-components";
global.alert = jest.fn();

import mockRouter from "../../mock/routerMock";

jest.mock("next/router", () => ({
  useRouter() {
    return mockRouter;
  },
}));
jest.spyOn(require("next/router"), "useRouter");

describe("Home Page", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("renders Home  page unchanged", () => {
    waitFor(() =>
      jest.spyOn(internalAxiosInstance, "get").mockResolvedValue({
        totalIncome: "10000",
        totolExpenses: "20000",
      })
    );
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
  it("should fetch summary data successfulty", () => {
    jest.spyOn(internalAxiosInstance, "get").mockResolvedValue({
      id: "mock-id",
      totalExpenses: 10000,
      totalIncome: 20000,
    });
    render(<Home />);
    waitFor(() => {
      expect(screen.getByText("Total Income: 20000")).toBeInTheDocument();
      expect(screen.getByText("Total Expenses: 10000")).toBeInTheDocument();
    });
  });
  it("should submit data and create transaction successful", () => {
    jest.spyOn(internalAxiosInstance, "post").mockResolvedValue({});
    const { getByTestId } = render(<Home />);
    const file = new File(["(⌐□_□)"], "chucknorris.png", { type: "image/png" });
    Object.defineProperty(file, "size", { value: 1024 * 1024 * 3 });

    const upload = getByTestId("upload-button");
    Object.defineProperty(upload, "files", {
      value: [file],
    });

    fireEvent.change(upload);
    fireEvent.input(screen.getByLabelText("Name"), {
      target: { value: "Mock Name" },
    });

    fireEvent.input(screen.getByLabelText("Amount"), {
      target: { value: 1000 },
    });
    fireEvent.click(screen.getByText("income"));
    fireEvent.click(screen.getByText("Record"));
  });
  it("should submit data and creat transaction not successful", () => {
    jest.spyOn(internalAxiosInstance, "post").mockRejectedValue("mock");
    const { getByTestId } = render(<Home />);
    const file = new File(["(⌐□_□)"], "chucknorris.png", { type: "image/png" });
    Object.defineProperty(file, "size", { value: 1024 * 1024 * 3 });
    waitFor(() => {
      expect(screen.getByText("Total Income: 20000")).toBeInTheDocument();
      expect(screen.getByText("Total Expenses: 10000")).toBeInTheDocument();
    });
    const upload = getByTestId("upload-button");
    Object.defineProperty(upload, "files", {
      value: [file],
    });

    fireEvent.change(upload);
    fireEvent.input(screen.getByLabelText("Name"), {
      target: { value: "Mock Name" },
    });

    fireEvent.input(screen.getByLabelText("Amount"), {
      target: { value: 1000 },
    });
    fireEvent.click(screen.getByText("income"));
    fireEvent.click(screen.getByText("Record"));
  });
  it("should have HOC calling", async () => {
    render(<HomePage />);
    waitFor(() => expect(Home).toBeCalled());
  });
});
