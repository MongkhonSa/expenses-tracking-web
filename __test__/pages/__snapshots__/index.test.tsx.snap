// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders Home  page unchanged 1`] = `
.c0 {
  max-width: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="ant-row ant-row-end"
  >
    <div
      class="ant-col "
    >
      <button
        class="ant-btn ant-btn-default"
        type="button"
      >
        <a
          href="/"
        >
          Home
        </a>
      </button>
    </div>
    <div
      class="ant-col "
    >
      <button
        class="ant-btn ant-btn-default"
        type="button"
      >
        <a
          href="/report"
        >
          Report
        </a>
      </button>
    </div>
    <div
      class="ant-col "
    >
      <button
        class="ant-btn ant-btn-default"
        type="button"
      >
        <h4>
          logout
        </h4>
      </button>
    </div>
  </div>
  <div
    class="ant-row ant-row-center"
  >
    <h1>
      Dashboard
    </h1>
     
  </div>
  <div
    class="ant-row ant-row-center"
  >
    <div
      class="ant-col ant-col-8"
    >
      <div
        class="ant-row"
      >
        Total Income: undefined
      </div>
      <div
        class="ant-row"
      >
        Total Expenses: undefined
      </div>
    </div>
    <div
      class="ant-col"
    >
      <form
        class="ant-form ant-form-horizontal c0"
      >
        <div
          class="ant-form-item"
        >
          <div
            class="ant-row ant-form-item-row"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="categoryName"
                title="Name"
              >
                Name
              </label>
            </div>
            <div
              class="ant-col ant-col-14 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input"
                    id="categoryName"
                    placeholder="Category name"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item"
        >
          <div
            class="ant-row ant-form-item-row"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="amount"
                title="Amount"
              >
                Amount
              </label>
            </div>
            <div
              class="ant-col ant-col-14 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <div
                    class="ant-input-number ant-input-number-in-form-item"
                  >
                    <div
                      class="ant-input-number-handler-wrap"
                    >
                      <span
                        aria-disabled="false"
                        aria-label="Increase Value"
                        class="ant-input-number-handler ant-input-number-handler-up"
                        role="button"
                        unselectable="on"
                      >
                        <span
                          aria-label="up"
                          class="anticon anticon-up ant-input-number-handler-up-inner"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="up"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                      <span
                        aria-disabled="false"
                        aria-label="Decrease Value"
                        class="ant-input-number-handler ant-input-number-handler-down"
                        role="button"
                        unselectable="on"
                      >
                        <span
                          aria-label="down"
                          class="anticon anticon-down ant-input-number-handler-down-inner"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="down"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                    <div
                      class="ant-input-number-input-wrap"
                    >
                      <input
                        aria-valuemin="0"
                        autocomplete="off"
                        class="ant-input-number-input"
                        id="amount"
                        placeholder="amount"
                        role="spinbutton"
                        step="1"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item"
        >
          <div
            class="ant-row ant-form-item-row"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="type"
                title="Type"
              >
                Type
              </label>
            </div>
            <div
              class="ant-col ant-col-14 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <div
                    class="ant-radio-group ant-radio-group-outline"
                    id="type"
                  >
                    <div
                      class="ant-row"
                    >
                      <label
                        class="ant-radio-wrapper ant-radio-wrapper-in-form-item"
                      >
                        <span
                          class="ant-radio"
                        >
                          <input
                            class="ant-radio-input"
                            type="radio"
                            value="income"
                          />
                          <span
                            class="ant-radio-inner"
                          />
                        </span>
                        <span>
                          income
                        </span>
                      </label>
                      <label
                        class="ant-radio-wrapper ant-radio-wrapper-in-form-item"
                      >
                        <span
                          class="ant-radio"
                        >
                          <input
                            class="ant-radio-input"
                            type="radio"
                            value="expenses"
                          />
                          <span
                            class="ant-radio-inner"
                          />
                        </span>
                        <span>
                          expenses
                        </span>
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item"
        >
          <div
            class="ant-row ant-form-item-row"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class=""
                for="file"
                title="Upload bill"
              >
                Upload bill
              </label>
            </div>
            <div
              class="ant-col ant-col-14 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <span
                    class=""
                  >
                    <div
                      class="ant-upload ant-upload-select ant-upload-select-text"
                    >
                      <span
                        class="ant-upload"
                        role="button"
                        tabindex="0"
                      >
                        <input
                          accept="image/*"
                          data-testid="upload-button"
                          id="file"
                          style="display: none;"
                          type="file"
                        />
                        <button
                          class="ant-btn ant-btn-default"
                          type="button"
                        >
                          <span
                            aria-label="upload"
                            class="anticon anticon-upload"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="upload"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                              />
                            </svg>
                          </span>
                          <span>
                            Click to Upload
                          </span>
                        </button>
                      </span>
                    </div>
                    <div
                      class="ant-upload-list ant-upload-list-text"
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item"
        >
          <div
            class="ant-row ant-form-item-row"
          >
            <div
              class="ant-col ant-col-16 ant-col-offset-4 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <button
                    class="ant-btn ant-btn-primary"
                    type="submit"
                  >
                    <span>
                      Record
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
